src/workerd/api/sockets.c++

```diff
@@ -104,16 +104,20 @@ jsg::Ref<Socket> setupSocket(
     fulfiller->fulfill(true);
   });
 
-  auto watchForDisconnectTask = connection->whenWriteDisconnected()
-      .then([&disconnectedFulfiller]() {
-    disconnectedFulfiller.fulfill(false);
-  }, [&disconnectedFulfiller](kj::Exception&& e) {
-    disconnectedFulfiller.reject(kj::mv(e));
-  }).eagerlyEvaluate(
-        [deferredCancelDisconnected = kj::mv(deferredCancelDisconnected)](kj::Exception&& e) {
-    // Should never get here because we caught exceptions above.
-    LOG_EXCEPTION("socketWatchForDisconnectTask", e);
-  });
+  static auto constexpr handleDisconnected =
+      [](kj::Promise<void> disconnected, auto& fulfiller) -> kj::Promise<void> {
+    try {
+      co_await disconnected;
+      fulfiller.fulfill(false);
+    } catch (...) {
+      auto exception = kj::getCaughtExceptionAsKj();
+      fulfiller.reject(kj::mv(exception));
+    }
+  };
+
+  auto watchForDisconnectTask = handleDisconnected(
+      connection->whenWriteDisconnected(),
+      disconnectedFulfiller).attach(kj::mv(deferredCancelDisconnected));
```

Comment:
I wonder if this .attach could be converted too?

Reply:
This attach is intended to be run if the returned `kj::Promise` is dropped (thereby canceling the coroutine). I think this one is best left as is, or if we do refactor this, then we can do so separately later.